Library/Homebrew/extend/cached.rb

```diff
@@ -0,0 +1,49 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Cached
+  module Clear
+    sig { void }
+    def clear_cache
+      super if defined?(super)
+
+      return unless defined?(@cached_method_calls)
+
+      remove_instance_variable(:@cached_method_calls)
+    end
+  end
+
+  sig { params(method: Symbol).returns(Symbol) }
+  def cached(method)
+    uncached_instance_method = instance_method(method)
+
+    define_method(method) do |*args, **options, &block|
+      @cached_method_calls ||= T.let({}, T.nilable(T::Hash[Symbol, T::Hash[T.untyped, T.untyped]]))
+      cache = @cached_method_calls[method] ||= {}
+
+      key = [args, options, block]
+      if cache.key?(key)
+        cache.fetch(key)
+      else
+        cache[key] = uncached_instance_method.bind(self).call(*args, **options, &block)
+      end
+    end
+  end
+
+  sig { params(method: Symbol).returns(Symbol) }
+  def cached_class_method(method)
+    uncached_singleton_method = singleton_method(method)
+
+    define_singleton_method(method) do |*args, **options, &block|
+      @cached_method_calls ||= T.let({}, T.nilable(T::Hash[Symbol, T::Hash[T.untyped, T.untyped]]))
+      cache = @cached_method_calls[method] ||= {}
+
+      key = [args, options, block]
+      if cache.key?(key)
+        cache[key]
+      else
+        cache[key] = uncached_singleton_method.call(*args, **options, &block)
+      end
+    end
+  end
```

Comment:
I'm not sure it really makes sense for cached methods to take arguments so this can probably be simplified a bit more. None of the cache methods in `Tap` take arguments right now. At the very least, I think it's a bad idea to allow cached methods to take blocks. I don't think it's be terribly difficult to add tests to make sure that methods are defined that way too.

Reply:
Also, I'm a bit surprised that you can use a block as a hash key. I guess I'd just never considered even attempting that.

Reply:
> I'm not sure it really makes sense for cached methods to take arguments so this can probably be simplified a bit more. None of the cache methods in `Tap` take arguments right now. At the very least, I think it's a bad idea to allow cached methods to take blocks.

Agreed.

Reply:
The block may not be needed though.

Reply:
Let's remove this functionality unless it's needed in 3+ places.

Reply:
There is also a pattern in the code to create complex hash keys by concatenating strings with dashes. I'm not sure if this is preferable though to just using arrays of objects. Does anyone have strong feelings either way?